keps/sig-network/4963-kube-proxy-flowtables-fastpath/README.md

```diff
@@ -0,0 +1,463 @@
+}
+```
+
+Kube-proxy will allow users to opt-in for Service acceleration, since it is not possible to guarantee that a feature that completely bypass the kernel has unintended consequences in some environments or incompatibility with other networking components.
+
+The opt-in behavior will be based on network interface names and attributes and will use CEL expressions, similar to how we use `--pod-interface-name-prefix` in [KEP-2450](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules). The powerful matching expression allow users to deal with more complex environments: nodes with multiple interfaces and only some of them with hardware offload, matches on alias to select only some specific kind of pods, ...
+
+Some example expressions are:
+
+- `interface.name.startsWith(\"eth\")` selects all interfaces with prefix `eth`
+- `interface.type == \"veth\"` selects all interfaces with type `veth`
+- `interface.alias == \"offload\"` selects all interfaces with alias set to `offload`
+
+This expressions can also be concatenated, see https://github.com/google/cel-go for more information.
+
+Kernel bypassing will only be applied once the connection is ESTABLISHED, the rationale behind this is that, once the connection is accelerated, it bypass the network stack, by waiting for the connection to be established we ensure the integration with other network applications (network policies, service meshes, ... ) in the node and we avoid possible performance problems or possible DOS attacks by trying to accelerate every single connection on a node.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to use Service traffic acceleration with kube-proxy to improve the performance of my applications.
+
+### Risks and Mitigations
+
+Once the network traffic moves to the datapath it completely bypass the kernel stack, so
+any other network applications that depend on the packets going through the network stack (monitoring per example) we'll not be able to see the connection data. The feature will only
+apply the fast path on established connections, since most of the network applications are statefuls, usually is safe to think that once a connection is established no additional operations are required on it.
+
+## Design Details
+
+This feature will only work with kube-proxy nftables mode.
+
+Users will be able to opt-in to Service traffic acceleration by passing a CEL expression using the flag `--accelerated-interface-expression` or the configuration option `AcceleratedInterfaceExpression` to match the network interfaces in the node that are subjet to Service traffic acceleration. The absence of a CEL expression disables the feature.
+
+Kube-proxy will create a `flowtable` in the kube-proxy table with the name `kube-proxy-flowtable` and will monitor the network interfaces in the node to populate the `flowtable` with the interfaces that match the configured CEL expression.
```

Comment:
At some point we need to investigate the performance implications of this, since it seems clear that this isn't really how flowtables were meant to be used, and there could potentially be things that are `O(n^2)` in the length of `devices`.

(Also, I wouldn't be surprised if it turned out that recreating the flowtable with an updated `devices` value discards all existing `flow offload` entries. This isn't _horrible_ since the offload will just get recreated on the next packet, but it's not _ideal_ and suggests maybe we should be working with the netfilter team to optimize this all a bit more...)

(This is maybe a topic for "Risks and Mitigations".)

Reply:
Good point, I also need to consult this with the netfilter folks

Reply:
> At some point we need to investigate the performance implications of this, since it seems clear that this isn't really how flowtables were meant to be used, and there could potentially be things that are `O(n^2)` in the length of `devices`.

Regarding "since it seems clear that this isn't really how flowtables were meant to be used", could you expand on that? I'm unsure what the intended use of flowtables is, and how this use case is unintended. 

I'm assuming this may be written down somewhere, so happy if you point me at a readme or similar

Reply:
It's not really written down, and I don't know exactly what the model is. But if they expected people to set up flows between every pair of devices on the node, then they wouldn't have required you to enumerate specific devices, right?

Reply:
> But if they expected people to set up flows between every pair of devices on the node, then they wouldn't have required you to enumerate specific devices, right?

AFAIK the flowtable (cache table) is per device, so this may be linking the specific flow to the specific device